### PR TITLE
Improving logic around visiting and flattening GroupJoin-SelectMany-DefaultIfEmpty queries.

### PR DESCRIPTION
**Pull request by maumar**
Previously, when we encountered this type of expression we would produce LEFT JOIN in the SelectExpression but at the same time rewrite the query model, pruning the GJ and DIE clauses, replacing it with regular join clause.
This is problematic, because QM no longer adequately represents the query that is being generated - if the new QM were visited again, we would produce INNER JOIN.

Fix is to not modify the QM, but instead keep track of additional from clauses that were flattened (and not visit them in the VisitAdditionalFromClause), as well as properly remap query/shapers to point to additional from clause, rather than the join.

### CODE REVIEW

**Comment by anpete** [comment]
Nice!

File: src/EFCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 1]
```diff
@@ -58,6 +58,7 @@ public class RelationalQueryModelVisitor : EntityQueryModelVisitor
         private ResultOperatorBase _groupResultOperatorInQueryModel;
 
         private readonly List<GroupJoinClause> _unflattenedGroupJoinClauses = new List<GroupJoinClause>();
+        private readonly List<AdditionalFromClause> _flattenedAdditionalFromClauses = new List<AdditionalFromClause>();
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryModelVisitor.cs]: We can easily get rid of this logic by matching AdditionalFromClause with _unflattenedGroupJoinClauses.

**Comment by maumar** [comment]
sqlite is broken

**Comment by smitpatel** [comment]
😞

**Comment by maumar** [comment]
fixed!

**Pull request merged**